Bulk memory support

rachel-bousfield opened the pull request:
Adds support for `MemoryFill` and `MemoryCopy` so that Nitro can be built with the latest Rust and Go compilers.

Additionally this PR
- Adds support for the latest Rust stable `1.68.2`
- Refactors hostio with a diff from Stylus
- Adds a bulk memory test

joshuacolvin0 reviewed (dismissed):
LGTM

joshuacolvin0 reviewed (dismissed):
LGTM

PlasmaPower commented on arbitrator/prover/src/bulk_memory.wat:
```diff
@@ -0,0 +1,148 @@
+            (then
+                ;; trap if out of bounds
+                (i64.gt_u
+                    ;; get the last source offset
```
This comment is slightly incorrect -- it's not the last source offset, it's that plus one. The logic itself is correct though.

rachel-bousfield replied:
Fixed!

PlasmaPower commented on arbitrator/prover/src/bulk_memory.wat:
```diff
@@ -0,0 +1,148 @@
+
+                ;; do the copy
+                (loop $forward
+                    ;; see if there's more data to set
```
For efficiency, we should first do a loop where we copy 64 bits at a time, until `size - offset < 8`. Then we enter this loop as-is.

rachel-bousfield replied:
Def agree for efficiency's sake. Originally I'd avoided doing this optimization just for simplicity and safety, but it's an 8x gain so it's worth doing. Btw I went with `offset + 8` to avoid an underflow but it's the same idea.

PlasmaPower commented on arbitrator/prover/src/bulk_memory.wat:
```diff
@@ -0,0 +1,148 @@
+            (else
+                ;; trap if out of bounds
+                (i64.gt_u
+                    ;; get the last destination offset
```
Again, the comment is slightly incorrect, but the logic is correct.

rachel-bousfield replied:
Fixed!

PlasmaPower commented on arbitrator/prover/src/bulk_memory.wat:
```diff
@@ -0,0 +1,148 @@
+                (if (then unreachable))
+
+                ;; do the copy
+                (loop $backward
```
Again, let's do an 64 bit loop first for efficiency.

rachel-bousfield replied:
See other :)

PlasmaPower commented on arbitrator/prover/src/bulk_memory.wat:
```diff
@@ -0,0 +1,148 @@
+                    local.get $offset
+                    i32.const 1
+                    i32.add
+                    local.set $offset
```
We should consider incrementing dest and decrementing size instead of using an offset. I think right now it would be less efficient, but I think the check in the 64 bit loop might be more efficient without a separate offset. I'm not sure it works out to an overall savings though.

PlasmaPower commented on arbitrator/prover/test-cases/bulk-memory.wat:
```diff
@@ -0,0 +1,54 @@
+        (memory.copy (i32.const 0x1009) (i32.const 0x100a) (i32.const 1))  ;; -888525--225-25-----------------
+        (memory.copy (i32.const 0x100f) (i32.const 0x1001) (i32.const 1))  ;; -888525--225-258----------------
+        (memory.copy (i32.const 0x100f) (i32.const 0x1000) (i32.const 32)) ;; ----------------888525--225-258-
+        (memory.copy (i32.const 0x1001) (i32.const 0x100f) (i32.const 32)) ;; --888525--225-258---------------
```
I think this does properly test both forward and backwards overlapping copies, but to make sure, could you replace the first check in memory_copy to determine if the copy should go backwards or forwards with an `i32.const 0` and `i32.const 1` and make sure this test fails in either case? Thanks

rachel-bousfield commented on arbitrator/prover/src/main.rs:
```diff
@@ -500,6 +500,7 @@ fn main() -> Result<()> {
     }
 
     if opts.require_success && mach.get_status() != MachineStatus::Finished {
+        println!("Machine didn't finish: {}", mach.get_status().red());
```
Added!

rachel-bousfield commented on arbitrator/prover/src/host.rs:
```diff
@@ -1,108 +1,224 @@
+        ("env", "wavm_read_pre_image")          => func!([I32, I32], [I32]),
+        ("env", "wavm_read_inbox_message")      => func!([I64, I32, I32], [I32]),
+        ("env", "wavm_read_delayed_inbox_message") => func!([I64, I32, I32], [I32]),
+        ("env", "wavm_halt_and_set_finished")      => func!(),
```
That makes sense, implemented!

It does have some minor awkwardness in their being duplicate names, but I've made `Hostio` use the corresponding `InternalFunc` when possible.

rachel-bousfield commented on arbitrator/prover/src/host.rs:
```diff
@@ -1,108 +1,224 @@
+    let mut funcs = vec![];
+
+    // order matters!
+    funcs.push(op_func(
```
I've added a couple to do this, anticipating what's coming in Stylus

rachel-bousfield commented on arbitrator/prover/src/host.rs:
```diff
@@ -1,108 +1,224 @@
+        let bin = binary::parse(&wasm).expect("failed to parse bulk_memory.wasm");
+        [0, 1].map(|i| {
+            let code = &bin.codes[i];
+            let ty = &bin.types[bin.functions[i] as usize];
```
Asserted!

rachel-bousfield commented:
I believe I've addressed all the comments. I went ahead and updated the Stylus side so we could run the full testsuite on the updated implementation. That's the reason you'll see a couple of exports added in my most recent commit :)

PlasmaPower reviewed (approved):
LGTM
